sys/net/gnrc/routing/ipv6_auto_subnets/gnrc_ipv6_auto_subnets.c

```diff
@@ -52,14 +64,130 @@
  * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
  */
 
+#include "net/gnrc/ipv6.h"
+#include "net/gnrc/netif.h"
+#include "net/gnrc/netif/hdr.h"
+#include "net/gnrc/udp.h"
 #include "net/gnrc/ipv6/nib.h"
 #include "net/gnrc/ndp.h"
+#include "random.h"
+#include "xtimer.h"
+
+/**
+ * @brief Port for the custom UDP sync protocol
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_PORT          (16179)
+#endif
+
+/**
+ * @brief Max number of other routers on the same link
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX     (4)
+#endif
+
+/**
+ * @brief How often the number subnets should be announced by the routers
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_PER_PERIOD
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_TX_PER_PERIOD (3)
+#endif
+/**
+ * @brief How long to wait for other routers annoucements before resending
+ *        or creating subnets when the retry counter is exhausted
+ */
+#ifndef CONFIG_GNRC_IPV6_AUTO_SUBNETS_TIMEOUT_MS
+#define CONFIG_GNRC_IPV6_AUTO_SUBNETS_TIMEOUT_MS    (50)
+#endif
+
+#define SERVER_THREAD_STACKSIZE                     (THREAD_STACKSIZE_DEFAULT)
+#define SERVER_MSG_QUEUE_SIZE                       (CONFIG_GNRC_IPV6_AUTO_SUBNETS_PEERS_MAX)
+#define SERVER_MSG_TYPE_TIMEOUT                     (0x8fae)
 
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
 static char addr_str[IPV6_ADDR_MAX_STR_LEN];
 
+#if !IS_USED(MODULE_GNRC_IPV6_AUTO_SUBNETS_SIMPLE)
+
+/**
+ * @brief Custom UDP sync protocol
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t version : 4;    /**< version number, should be 0 */
+    uint8_t flags   : 4;    /**< flags, unused, should be 0 */
```

Comment:
We don't like bit fields because of reasons (honestly can't why remember, but that's why I used flags and bit masks, e.g., in the GNRC headers).

Reply:
Also, do you mean “should” or “must”?

Reply:
And the flags are “reserved for future use”, not unused, right?

Reply:
> We don't like bit fields because of reasons

But bit fields are a perfectly cromulent C feature, what's wrong with them?

Reply:
I think, and again, I don't really remember exactly, there were two problems:

1. They are not compatible with C++. Granted, not a problem here, since this is a `.c` file.
2. They result in larger binaries compared to doing it by hand.

As 1. is invalid here and 2. is maybe a somewhat premature optimization for something that is internal to this module, I agree (if those reasons are true) that we can keep them. Maybe @kaspar030 or @haukepetersen remember?

Reply:
There's a third:
3. bitfield ordering is implementation defined:

```
C99 6.7.2.1-11:An implementation may allocate any addressable storage unit large enough to hold a bit- field. If enough space remains, a bit-field that immediately follows another bit-field in a structure shall be packed into adjacent bits of the same unit. If insufficient space remains, whether a bit-field that does not fit is put into the next unit or overlaps adjacent units is implementation-defined. The order of allocation of bit-fields within a unit (high-order to low-order or low-order to high-order) is implementation-defined. The alignment of the addressable storage unit is unspecified.
```

Reply:
Ok, so bad for networking headers, such as this one...

Reply:
this is wire data, right? compile it with a different compiler, the ordering might change.

Reply:
> this is wire data, right? compile it with a different compiler, the ordering might change.

Yep.

Reply:
So for the sake of a lot of headache spared, when nodes later do not communicate properly, please use flags, @benpicco or if that is too much hassle use an 8-bit version number ;-P. It was me who proposed to use a nibble after all.

Reply:
> **If enough space remains**, a bit-field that immediately follows another bit-field in a structure shall be **packed into adjacent bits of the same unit**. If **insufficient space remains**, whether a bit-field that does not fit is put into the next unit or overlaps adjacent units is implementation-defined.

This is only an issue if you write

```C
struct foo {
    uint8_t version : 4; 
    uint8_t flags   : 6;
} _auto_subnets_request_t;
```

Now `flags` could either use some bits of the first byte or entirely occupy the second byte.

If the bits add up to a whole byte, there should be no ambiguity.